### PR TITLE
Added the ability to send DMs from bot accounts

### PR DESCRIPTION
**Pull request by icelander**
#### Summary

A customer had the requirement to send a DM from a bot account, so I added the ability. Any channel name that starts with an `@` symbol that is also a bot user will have a DM sent with the message configured in the Action's `ActionDirectMessagePost`

### CODE REVIEW

File: server/manifest.go [hunk 1]
```diff
@@ -7,5 +7,5 @@ var manifest = struct {
 	Version string
 }{
 	ID:      "com.mattermost.welcomebot",
-	Version: "1.2.0",
```
Comment (jfrerich) [inline server/manifest.go:10]: Let's leave this out for now. The version will be changed in its own PR after numerous PRs have been merged.

File: server/welcomebot.go [hunk 2]
```diff
@@ -228,12 +230,55 @@ func (p *Plugin) processActionMessage(messageTemplate MessageTemplate, action *A
...
+			p.API.LogError("Specified DM user is not a bot, continuing to next channel", "channelName", channelName)
+			return
+		}
+		
+		dmChannel, err := p.API.GetDirectChannel(dmUser.Id, action.Context.UserID)
+
+		if err != nil {
+			p.API.LogError("Couldn't create or get DM channel, continuing to next channel", "user_id", action.Context.UserID, "channelName", channelName, "channel_id", dmChannel.Id)
+			return
+		}
```
Comment (jfrerich) [inline server/welcomebot.go:253]:
```suggestion
		dmChannel, err := p.API.GetDirectChannel(dmUser.Id, action.Context.UserID)
		if err != nil {
			p.API.LogError("Couldn't create or get DM channel, continuing to next channel", "user_id", action.Context.UserID, "channelName", channelName, "channel_id", dmChannel.Id)
			return
		}
```

File: server/welcomebot.go [hunk 3]
```diff
@@ -228,12 +230,55 @@ func (p *Plugin) processActionMessage(messageTemplate MessageTemplate, action *A
 }
 
 func (p *Plugin) joinChannel(action *Action, channelName string) {
```
Comment (jfrerich) [inline server/welcomebot.go:232]: I'd prefer to have this function split up a bit.  Something like the following: 
```go 
	// If it begins with @ create a DM channel
	if strings.HasPrefix(channelName, "@") {
       doSomethingForDMs()
       return
    } 

    // Otherwise treat it like a normal channel
    if channel, err := p.API.GetChannelByName(action.Context.TeamID, channelName, false); err == nil {
		if _, err := p.API.AddChannelMember(channel.Id, action.Context.UserID); err != nil {
			p.API.LogError("Couldn't add user to the channel, continuing to next channel", "channel_name", channelName, "user_id", action.Context.UserID, channel.Id)
			return
		}
	} else {
		p.API.LogError("failed to get channel, continuing to the next channel", "channel_name", channelName, "user_id", action.Context.UserID)
		return
	}
```

Comment (jfrerich) [reply server/welcomebot.go:232]: This should make the code a bit cleaner and the PR will only show changes that are specific to the new functionality because the diff won't be changed for the existing, unchanged code because of formatting.

**Review by jfrerich — CHANGES_REQUESTED** [review]
Thanks @icelander! I have a couple of suggestions :)

File: server/welcomebot.go (hunk 2 above)
Comment (mickmister) [reply server/welcomebot.go:253]: What is being suggested here? Removing the newline at line 249?

File: server/welcomebot.go [hunk 4]
```diff
@@ -228,12 +230,55 @@ func (p *Plugin) processActionMessage(messageTemplate MessageTemplate, action *A
...
-	if channel, err := p.API.GetChannelByName(action.Context.TeamID, channelName, false); err == nil {
-		if _, err := p.API.AddChannelMember(channel.Id, action.Context.UserID); err != nil {
-			p.API.LogError("Couldn't add user to the channel, continuing to next channel", "user_id", action.Context.UserID, "channel_id", channel.Id)
+	// If it begins with @ create a DM channel
+	if strings.HasPrefix(channelName, "@") {
+		r := []rune(channelName)
+		dmUser, userErr := p.API.GetUserByUsername(string(r[1:]))
```
Comment (mickmister) [inline server/welcomebot.go:236]: You can perform the same slicing on strings as you can with arrays:

```suggestion
		dmUser, userErr := p.API.GetUserByUsername(channelName[1:])
```

File: server/welcomebot.go (hunk 3 above)
Comment (mickmister) [reply server/welcomebot.go:232]: On top of this, I recommend having the function `doSomethingForDMs` return an error, and have the caller log the error. To include extra information when returning an error, you can use `errors.Wrap` or `errors.Wrapf`:

```go
err := doSomethingForDMs(...)
if err != nil {
    p.API.LogError("failed to handle DM channel, continuing to next channel. " + err.Error())
}

// Then in doSomethingForDMs()
func doSomethingForDMs(username string, ...) error {
    dmUser, userErr := p.API.GetUserByUsername(username)
    if userErr != nil {
        return errors.Wrapf(userErr, "couldn't find DM channel for username %s", username)
    }
    // ...
}
```

Doing this decouples the error creation and logging functionality

**Review by mickmister — CHANGES_REQUESTED** [review]
Nice work @icelander :+1: Implementation of the feature looks good

I have a request that goes along with Jason's comment above, to have the extracted function return an error so there is only one log statement for the error

**Comment by Kshitij-Katiyar** [comment]
Closing the PR due to the PR https://github.com/mattermost/mattermost-plugin-welcomebot/pull/129
